pilot/pkg/model/endpointshards.go

```diff
@@ -288,44 +288,48 @@ func (e *EndpointIndex) UpdateServiceEndpoints(
 	newIstioEndpoints := istioEndpoints
 	if features.SendUnhealthyEndpoints.Load() {
 		oldIstioEndpoints := ep.Shards[shard]
-		newIstioEndpoints = make([]*IstioEndpoint, 0, len(istioEndpoints))
-
-		// Check if new Endpoints are ready to be pushed. This check
-		// will ensure that if a new pod comes with a non ready endpoint,
-		// we do not unnecessarily push that config to Envoy.
-		// Please note that address is not a unique key. So this may not accurately
-		// identify based on health status and push too many times - which is ok since its an optimization.
-		emap := make(map[string]*IstioEndpoint, len(oldIstioEndpoints))
-		nmap := make(map[string]*IstioEndpoint, len(newIstioEndpoints))
-		// Add new endpoints only if they are ever ready once to shards
-		// so that full push does not send them from shards.
-		for _, oie := range oldIstioEndpoints {
-			emap[oie.Address] = oie
-		}
-		for _, nie := range istioEndpoints {
-			nmap[nie.Address] = nie
-		}
 		needPush := false
-		for _, nie := range istioEndpoints {
-			if oie, exists := emap[nie.Address]; exists {
-				// If endpoint exists already, we should push if it's health status changes.
-				if oie.HealthStatus != nie.HealthStatus {
+		if oldIstioEndpoints == nil {
+			// If there are no old endpoints, we should push with incoming endpoints as there is nothing to compare.
+			needPush = true
+		} else {
+			newIstioEndpoints = make([]*IstioEndpoint, 0, len(istioEndpoints))
+			// Check if new Endpoints are ready to be pushed. This check
+			// will ensure that if a new pod comes with a non ready endpoint,
+			// we do not unnecessarily push that config to Envoy.
+			// Please note that address is not a unique key. So this may not accurately
+			// identify based on health status and push too many times - which is ok since its an optimization.
+			emap := make(map[string]*IstioEndpoint, len(oldIstioEndpoints))
+			nmap := make(map[string]*IstioEndpoint, len(newIstioEndpoints))
+			// Add new endpoints only if they are ever ready once to shards
+			// so that full push does not send them from shards.
+			for _, oie := range oldIstioEndpoints {
+				emap[oie.Address] = oie
+			}
+			for _, nie := range istioEndpoints {
+				nmap[nie.Address] = nie
+			}
+			for _, nie := range istioEndpoints {
+				if oie, exists := emap[nie.Address]; exists {
+					// If endpoint exists already, we should push if it's health status changes.
+					if oie.HealthStatus != nie.HealthStatus {
+						needPush = true
+					}
+					newIstioEndpoints = append(newIstioEndpoints, nie)
+				} else {
```

Comment:
Can we use the old state when available? Or an option to configure that would be desireable.

I understand choosing the option may introduce inconsistent behaviour from different istiod pods.

Reply:
I think it is not desirable to have inconsistent behaviour. The reason why we disabled panic threshold is it leads to the case where a non ready pod gets request which is against k8s.